File: app/src/main/java/com/pleon/buyt/database/repository/AddItemRepository.java
package com.pleon.buyt.database.repository;

import android.app.Application;
import android.os.AsyncTask;

import com.pleon.buyt.database.AppDatabase;
import com.pleon.buyt.database.dao.ItemDao;
import com.pleon.buyt.database.dao.PurchaseDao;
import com.pleon.buyt.model.Item;
import com.pleon.buyt.model.Purchase;
import com.pleon.buyt.model.Store;

import java.util.Date;

import androidx.lifecycle.LiveData;

public class AddItemRepository {

    private ItemDao itemDao;
    private PurchaseDao purchaseDao;
    private LiveData<String[]> itemNames;

    public AddItemRepository(Application application) {
        itemDao = AppDatabase.getDatabase(application).itemDao();
        purchaseDao = AppDatabase.getDatabase(application).purchaseDao();
        itemNames = itemDao.getItemNames();
    }

    public void addItem(Item item) {
        new AddItemTask(itemDao).execute(item);
    }

    public void addPurchasedItem(Item item, Store store, Date purchaseDate) {
        new AddPurchasedItemTask(itemDao, purchaseDao, store, purchaseDate).execute(item);
    }

    public LiveData<String[]> getItemNames() {
        return itemNames;
    }

    private static class AddItemTask extends AsyncTask<Item, Void, Void> {

        private ItemDao itemDao;

        AddItemTask(ItemDao itemDao) {
            this.itemDao = itemDao;
        }

        @Override
        protected Void doInBackground(Item... items) {
            itemDao.insert(items[0]);
            return null;
        }
    }

    private static class AddPurchasedItemTask extends AsyncTask<Item, Void, Void> {

        private ItemDao itemDao;
        private PurchaseDao purchaseDao;
        private Store store;
        private Date date;

        AddPurchasedItemTask(ItemDao itemDao, PurchaseDao purchaseDao, Store store, Date date) {
            this.itemDao = itemDao;
            this.purchaseDao = purchaseDao;
            this.store = store;
            this.date = date;
        }

        @Override
        protected Void doInBackground(Item... items) {
            Purchase purchase = new Purchase(store.getStoreId(), date);
            long purchaseId = purchaseDao.insert(purchase);

            items[0].setPurchaseId(purchaseId);
            itemDao.insert(items[0]);

            return null;
        }
    }
}

File: app/src/main/kotlin/com/pleon/buyt/database/repository/AddItemRepository.kt
package com.pleon.buyt.database.repository

import android.app.Application
import com.pleon.buyt.database.getDatabase
import com.pleon.buyt.model.Item
import com.pleon.buyt.model.Purchase
import com.pleon.buyt.model.Store
import org.jetbrains.anko.doAsync
import java.util.*

class AddItemRepository(application: Application) {

    private val itemDao = getDatabase(application).itemDao()
    private val purchaseDao = getDatabase(application).purchaseDao()
    val itemNames = itemDao.getItemNames()

    fun addItem(item: Item) = doAsync { itemDao.insertItem(item) }

    fun addPurchasedItem(item: Item, store: Store, purchaseDate: Date) {
        doAsync {
            val purchase = Purchase(store.storeId, purchaseDate)
            purchaseDao.insert(purchase).also { item.purchaseId = it }
            itemDao.insertItem(item)
        }
    }
}
